File: algorithm/problem/tree/113. 路径总和 II.go
package tree

/*
给你二叉树的根节点 root 和一个整数目标和 targetSum ，找出所有 从根节点到叶子节点 路径总和等于给定目标和的路径。

叶子节点 是指没有子节点的节点。

输入：root = [5,4,8,11,null,13,4,7,2,null,null,5,1], targetSum = 22
输出：[[5,4,11,2],[5,8,4,5]]
*/

/*
解题思路:
	使用一个二维数组的地址透传 然后使用一个数组存储当前阶段访问过的元素
	注意 在使用前先拷贝数组
	避免 同数组在递归中出现混乱
*/

func pathSum(root *TreeNode, targetSum int) [][]int {
	total := make([][]int, 0)
	current := make([]int, 0)
	check(root, targetSum, &total, current)
	return total
}

func check(root *TreeNode, targetSum int, total *[][]int, current []int) {
	// 此处注意
	current1 := make([]int, 0)
	current1 = append(current1, current...)
	if root == nil {
		return
	}
	if root.Left == nil && root.Right == nil && targetSum-root.Val == 0 {
		current1 = append(current1, root.Val)
		*total = append(*total, current1)
		return
	}

	current1 = append(current1, root.Val)
	if root.Left != nil {
		check(root.Left, targetSum-root.Val, total, current1)
	}
	if root.Right != nil {
		check(root.Right, targetSum-root.Val, total, current1)
	}
}

File: algorithm/problem/tree/113. 路径总和 II.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> pathSum(TreeNode* root, int targetSum) {
        vector<vector<int>> init;
        vector<int> current;
        check(root, targetSum, init, current);
        return init;
    }

    void check(TreeNode* root, int targetSum, vector<vector<int>>& total, vector<int> current)
    {
        // 创建一个数组 将引用的值取出
        vector<int> newAarray(current);
        if (!root)
        {
            return;
        }
        if (!root->left && !root->right && targetSum - root->val == 0)
        {
            newAarray.push_back(root->val);
            total.push_back(newAarray);
            return;
        }
        newAarray.push_back(root->val);
        // 处理左子树
        if (root->left)
        {
            check(root->left, targetSum-root->val, total, newAarray);
        }

        // 处理右子树
        if (root->right)
        {
            check(root->right, targetSum-root->val, total, newAarray);
        }

        return;
    }
};

/*
注意:
    total需要改变需要传递引用
    current防止左右子树同时修改 进行值传递
*/